### PR TITLE
Client sync

### PR DESCRIPTION
**Pull request by ericfonseca**

### CODE REVIEW

File: web/src/app/models/game.ts [hunk 1]
```diff
@@ -93,4 +93,48 @@ export class Game {
       const answer = new Entry({ tiles, clue });
     });
   }
+
+  private printRow(row: Tile[]): string {
```
Comment (vqdo) [inline web/src/app/models/game.ts]: serializeRow

File: web/src/app/models/game.ts [hunk 2]
```diff
@@ -93,4 +93,48 @@ export class Game {
...
+    });
+    return rowStr;
+  }
+
+  public serializeBoard(): string {
+    let boardStr = '';
+    this.board.forEach((row: Tile[]) => {
+      boardStr += this.printRow(row);
+    });
+    return boardStr;
```
Comment (vqdo) [inline web/src/app/models/game.ts]: Use .map.

`this.board.map(() => this.printRow(row)).join('')`

File: web/src/app/models/game.ts [hunk 3]
```diff
@@ -93,4 +93,48 @@ export class Game {
       const answer = new Entry({ tiles, clue });
     });
   }
+
+  private printRow(row: Tile[]): string {
+    let rowStr = '';
+    row.forEach((tile: Tile) => {
```
Comment (vqdo) [inline web/src/app/models/game.ts]: Use .map:

```
row.forEach((tile) => {
  let rowStr;
  if (tile.value === -1) {
    rowStr = '#'; // or return '#';
  }
...
  return rowStr;
}).join('');
```

File: web/src/app/models/game.ts [hunk 4]
```diff
@@ -93,4 +93,48 @@ export class Game {
...
+  }
+
+  public deserializeBoard(boardStr: string) {
+    if (boardStr.length !== this.crossword.height * this.crossword.width) {
+      throw Error('incorrectly formatted board string');
+    }
+
+    for (let i = 0; i < boardStr.length; i++) {
+      const str = boardStr.charAt(i);
+      if (str !== ' ' && str !== '#') {
```
Comment (vqdo) [inline web/src/app/models/game.ts]: Should we set this to -1 if str === '#'

File: web/src/app/models/game.ts [hunk 5]
```diff
@@ -93,4 +93,48 @@ export class Game {
...
+    });
+    return boardStr;
+  }
+
+  public deserializeBoard(boardStr: string) {
+    if (boardStr.length !== this.crossword.height * this.crossword.width) {
+      throw Error('incorrectly formatted board string');
+    }
+
+    for (let i = 0; i < boardStr.length; i++) {
```
Comment (vqdo) [inline web/src/app/models/game.ts]: I prefer forEach. `boardStr.split('').forEach((ch) => { .... }`

File: web/src/app/services/crossword-data.service.ts [hunk 6]
```diff
@@ -60,23 +60,36 @@ export class CrosswordDataService {
...
       .mergeMap((res: GameResponse) => {
         return this.http.get(`${AppConfig.GAME_URL}/${gameId}/clues`)
           .map((clueRes: CluesResponse) => this.extractClues(clueRes))
           .map((clues: Clue[]) => {
-            return new Game({
+            const game = new Game({
               id: gameId,
               crossword: this.extractCrossword(res, clues),
             });
+            game.deserializeBoard(res.board_state);
```
Comment (vqdo) [inline web/src/app/services/crossword-data.service.ts]: We probably need to add an optional Game parameter (boardState) to the ctor.

File: web/src/app/services/crossword-data.service.ts [hunk 7]
```diff
@@ -60,23 +60,36 @@ export class CrosswordDataService {
...
+    const data = {
+      'board_state': boardStr,
+    };
+    return this.http.post(`${AppConfig.GAME_URL}/${game.id}`, JSON.stringify(data));
+  }
+
+  public sync(game: Game): Observable<Game> {
+    return this.uploadGame(game)
+      .mergeMap(() => {
+        return this.getGame(game.id);
```
Comment (vqdo) [inline web/src/app/services/crossword-data.service.ts]: This seems ok for now, but I expected the /<gameId> endpoint to return the same contents as getGame().

File: web/src/app/services/crossword-data.service.ts (hunk 6 above)
Comment (vqdo) [reply web/src/app/services/crossword-data.service.ts]: don't need to do this now

File: web/src/app/services/crossword-data.service.ts (hunk 7 above)
Comment (ericfonseca) [reply web/src/app/services/crossword-data.service.ts]: will be the next PR

File: web/src/app/models/game.ts (hunk 1 above)
Comment (ericfonseca) [reply web/src/app/models/game.ts]: agreed

File: web/src/app/models/game.ts (hunk 2 above)
Comment (ericfonseca) [reply web/src/app/models/game.ts]: agreed

File: web/src/app/models/game.ts (hunk 4 above)
Comment (ericfonseca) [reply web/src/app/models/game.ts]: uh i dont really agree, we send state before we receive it back. we should trust the client side state.

File: web/src/app/services/crossword-data.service.ts (hunk 6 above)
Comment (ericfonseca) [reply web/src/app/services/crossword-data.service.ts]: agree

File: web/src/app/models/game.ts (hunk 5 above)
Comment (ericfonseca) [reply web/src/app/models/game.ts]: actually we need the index to figure out what the row num, will leave for now

**Pull request merged**